### PR TITLE
Improvements to the ML-DSA code.

### PR DESCRIPTION
**Pull request by xvzcf**

### CODE REVIEW

File: libcrux-sha3/src/lib.rs [hunk 1]
```diff
@@ -237,12 +237,13 @@ pub mod portable {
...
         use generic_keccak::{
-            absorb_final, squeeze_first_five_blocks, squeeze_first_three_blocks, squeeze_next_block,
+            absorb_final, squeeze_first_block, squeeze_first_five_blocks,
+            squeeze_first_three_blocks, squeeze_next_block,
         };
 
         use super::*;
 
-        /// Initialise the SHAKE state.
+        /// Crate a new SHAKE-128 state object.
```
Comment (franziskuskiefer) [inline libcrux-sha3/src/lib.rs:246]:
```suggestion
        /// Create a new SHAKE-128 state object.
```

File: libcrux-sha3/src/lib.rs [hunk 2]
```diff
@@ -273,6 +274,31 @@ pub mod portable {
         pub fn shake128_squeeze_next_block(s: &mut KeccakState, out0: &mut [u8]) {
             squeeze_next_block::<1, u64, 168>(&mut s.state, [out0])
         }
+
+        /// Crate a new SHAKE-256 state object.
```
Comment (franziskuskiefer) [inline libcrux-sha3/src/lib.rs:278]:
```suggestion
        /// Create a new SHAKE-256 state object.
```

File: libcrux-sha3/src/lib.rs [hunk 3]
```diff
@@ -273,6 +274,31 @@ pub mod portable {
...
+        #[inline(always)]
+        pub fn shake256_init() -> KeccakState1 {
```
Comment (franziskuskiefer) [inline libcrux-sha3/src/lib.rs:280]: I think `KeccakState1` is `KeccakState` now.

File: libcrux-ml-dsa/tests/wycheproof_sign.rs [hunk 4]
```diff
@@ -50,7 +50,10 @@ macro_rules! wycheproof_sign_test {
                             signature.0.as_slice(),
                             hex::decode(test.sig).unwrap().as_slice()
                         );
-                    } // TODO: else, how should invalid signatures be handled?
+                    }
+                    // TODO: else, the generated signature is invalid; we can
```
Comment (franziskuskiefer) [inline libcrux-ml-dsa/tests/wycheproof_sign.rs:54]: Can you file an issue to track this?

File: libcrux-ml-dsa/src/sample.rs [hunk 5]
```diff
@@ -204,48 +207,60 @@ pub(crate) fn sample_mask_vector<const DIMENSION: usize, const GAMMA1_EXPONENT:
...
+                result.coefficients[sample_at] = 1 - 2 * ((*signs & 1) as i32);
+                *signs >>= 1;
+            }
 
-    let mut signs: u64 = 0;
-    for i in 0..8 {
-        signs |= (randomness.next().unwrap() as u64) << (8 * i);
+            if *out_index == result.coefficients.len() {
+                done = true;
+            }
```
Comment (franziskuskiefer) [inline libcrux-ml-dsa/src/sample.rs:231]: What about?

```suggestion
            done = *out_index == result.coefficients.len();
```

File: libcrux-ml-dsa/src/sample.rs [hunk 6]
```diff
@@ -143,17 +143,17 @@ pub(crate) fn rejection_sample_less_than_eta<const ETA: usize>(
 #[allow(non_snake_case)]
 #[inline(always)]
 fn sample_error_ring_element<const ETA: usize>(seed: [u8; 66]) -> PolynomialRingElement {
```
Comment (franziskuskiefer) [inline libcrux-ml-dsa/src/sample.rs:145]: This looks very similar to `sample_challenge_ring_element`. Can we re-use code?

File: libcrux-ml-dsa/src/sample.rs [hunk 7]
```diff
@@ -21,12 +21,12 @@ fn rejection_sample_less_than_field_modulus(
 
             let potential_coefficient = ((b2 << 16) | (b1 << 8) | b0) & 0x00_7F_FF_FF;
 
-            if potential_coefficient < FIELD_MODULUS && *sampled < COEFFICIENTS_IN_RING_ELEMENT {
+            if potential_coefficient < FIELD_MODULUS && *sampled < out.coefficients.len() {
```
Comment (franziskuskiefer) [inline libcrux-ml-dsa/src/sample.rs:24]: Why stop using the constant here?

File: libcrux-ml-dsa/src/ml_dsa_generic.rs [hunk 8]
```diff
@@ -270,16 +261,16 @@ pub(crate) fn sign<
 
     let mut attempt = 0;
 
+    // TODO: This loop will have to be rewritten to go through hax.
```
Comment (franziskuskiefer) [inline libcrux-ml-dsa/src/ml_dsa_generic.rs:264]: Can you collect these things in an issue for the verification?

**Review by franziskuskiefer — CHANGES_REQUESTED** [review]
lgtm, just a few nits and questions.

File: libcrux-ml-dsa/src/sample.rs (hunk 7 above)
Comment (xvzcf) [reply libcrux-ml-dsa/src/sample.rs:24]: It just seemed a bit more DRY, plus we don't have to import the constant in anymore, not a big deal though I can revert it if need be.

Comment (franziskuskiefer) [reply libcrux-ml-dsa/src/sample.rs:24]: No I think this is fine. The compiler knows this is a constant.

File: libcrux-ml-dsa/src/sample.rs (hunk 6 above)
Comment (xvzcf) [reply libcrux-ml-dsa/src/sample.rs:145]: I'm not sure, the basic structure is indeed similar (squeeze some bytes, pass these bytes to a function, and repeat if need be); but I think thematically they're different, `sample_error_ring_element` starts at 0 and systematically fills the ring element, whereas `sample_challenge_ring_element` is sprinkling a fixed number of positive or negative 1's randomly across the array.

File: libcrux-ml-dsa/tests/wycheproof_sign.rs (hunk 4 above)
Comment (xvzcf) [reply libcrux-ml-dsa/tests/wycheproof_sign.rs:54]: Opened https://github.com/cryspen/libcrux/issues/340 and made a reference to it in the comments.

File: libcrux-ml-dsa/src/ml_dsa_generic.rs (hunk 8 above)
Comment (xvzcf) [reply libcrux-ml-dsa/src/ml_dsa_generic.rs]: Opened https://github.com/cryspen/libcrux/issues/341 and referenced it in the comments.

**Review by franziskuskiefer — APPROVED** [review]

**Pull request merged**